### PR TITLE
linux: editor lifecycle improvements

### LINKED ISSUE #373: Auto save while typing

### LINKED ISSUE #381: Prompt the user when closing a modified, unsaved file
When a user closes a file that has been edited but _not_ saved, ask if they would like to save or close without saving.

This should be done after autosave. If autosave is enabled, the file might be saved automatically when closing each time.

### PR DESCRIPTION
**Pull request by Parth**
+ [x] Change window title based on open file
+ [x] Be able to detect when a file is dirty
+ [x] Add a `+` or `*` to the window title if the current file is dirty
+ [x] Warn prior to dirty file changes (fixes #381)
+ [x] Create setting to auto save dirty files after a configurable delay, default to auto save after 2 seconds of being idle
+ [x] Auto save dirty files (fixes #373)

### CODE REVIEW

**Review by raayan — APPROVED** [review]

File: clients/linux/src/app.rs [hunk 1]
```diff
@@ -296,11 +306,25 @@ impl LbApp {
     }
 
     fn open_file(&self, maybe_id: Option<Uuid>) -> LbResult<()> {
+        // Ask the user how the want to deal with unsaved changes
+        if self.state.borrow().open_file_dirty {
+            if let Some(open_file) = self.state.borrow().opened_file.clone() {
+                if !self.save_file_with_dialog(&open_file) {
+                    // File was not dealt with (dialog was closed) return early or we will lose unsaved work
+                    // Re-select the file that was open to make it clear the user's action was cancelled
+                    self.gui.account.sidebar.tree.select(&open_file.id);
```
Comment (steverusso) [inline clients/linux/src/app.rs:315]: The getter for `tree` is used 6 lines later. If we are going to the no getter route, this should probably be taken care of in an immediate followup PR.

File: clients/linux/src/auto_save.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    fn current_time() -> u128 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_millis()
+    }
+
+    pub fn auto_save_loop(state: Arc<Mutex<AutoSaveState>>) {
```
Comment (steverusso) [inline clients/linux/src/auto_save.rs:32]: `Arc<Mutex<Self>>` clearer?

**Review by steverusso — CHANGES_REQUESTED** [review]

File: clients/linux/src/auto_save.rs (hunk 2 above)
Comment (Parth) [reply clients/linux/src/auto_save.rs]: 463653a

File: clients/linux/src/app.rs (hunk 1 above)
Comment (Parth) [reply clients/linux/src/app.rs:315]: did this in this PR, in a future pr I can get rid of the other getters: 2f30396

**Review by steverusso — APPROVED** [review]

**Pull request merged**